services/data_plane_manager/src/main/java/com/futurewei/alcor/dataplane/utils/GoalStateHelper.java

```diff
@@ -0,0 +1,332 @@
+    // corner case check
+    if (networkConfiguration.getNeighborTable() == null
+        && networkConfiguration.getNeighborInfos() == null) return;
+    if (currentGroupHostIp.equals(ipHostIpMap.get(ip))
+        && networkConfiguration.getNeighborInfos().size()
+            == networkConfiguration.getNeighborTable().size()
+        && networkConfiguration.getNeighborTable().size() > 1) {
+      Set<String> hostIps = new HashSet();
+      for (NeighborEntry n : networkConfiguration.getNeighborTable()) {
+        hostIps.add(ipHostIpMap.get(n.getNeighborIp()));
+      }
+      if (hostIps.size() == 1)
+        for (InternalPortEntity i : portIdPortMap.values()) {
+          for (PortEntity.FixedIp fixedIp : i.getFixedIps()) {
+            if (fixedIp.getIpAddress().equals(ip)) return;
+          }
+        }
+    }
+    // construct the L3/L2 neighbor
+    if (neighborStates.containsKey(ip + "#" + Neighbor.NeighborType.L3)) return;
+    else if ((neighborStates.containsKey(ip + "#" + Neighbor.NeighborType.L2))
+        && (neighborType.equals(Neighbor.NeighborType.L3)))
+      neighborStates.remove(ip + "#" + Neighbor.NeighborType.L2);
+    Neighbor.NeighborConfiguration.FixedIp fixedIp =
+        Neighbor.NeighborConfiguration.FixedIp.newBuilder()
+            .setIpAddress(ip)
+            .setSubnetId(ipSubnetIdMap.get(ip))
+            .setNeighborType(neighborType)
+            .build();
+    Neighbor.NeighborConfiguration neighborConfiguration =
+        Neighbor.NeighborConfiguration.newBuilder()
+            .addFixedIps(fixedIp)
+            .setHostIpAddress(ipHostIpMap.get(ip))
+            .setMacAddress(ipMacMap.get(ip))
+            .setFormatVersion(GoalStateManager.FORMAT_REVISION_NUMBER)
+            .setRevisionNumber(Port.PortConfiguration.REVISION_NUMBER_FIELD_NUMBER)
+            .setId(ipPortIdMap.get(ip))
+            .setProjectId(networkConfiguration.getVpcs().get(0).getProjectId())
+            .setVpcId(networkConfiguration.getVpcs().get(0).getId())
+            .build();
+    Common.OperationType target =
+        getOperationType(
+            networkConfiguration.getOpType().equals(Common.OperationType.CREATE),
+            networkConfiguration.getOpType().equals(Common.OperationType.DELETE),
+            Common.OperationType.DELETE,
+            networkConfiguration.getOpType().equals(Common.OperationType.UPDATE),
+            Common.OperationType.UPDATE,
+            networkConfiguration.getOpType().equals(Common.OperationType.INFO),
+            Common.OperationType.INFO);
+    Neighbor.NeighborState neighborState =
+        Neighbor.NeighborState.newBuilder()
+            .setConfiguration(neighborConfiguration)
+            .setOperationType(target)
+            .build();
+    neighborStates.put(ip + "#" + neighborType, neighborState);
+    brandNewIps.add(ip);
+  }
+
+  @Nullable
+  Common.OperationType getOperationType(
+      boolean equals,
+      boolean equals2,
+      Common.OperationType delete,
```

Comment:
Does the parameter Common.Opeartion delete always means "resource deletion" or has different meaning in different usage?

Reply:
yes, first